Add script to manage windows services (and stuff)

bwatters-r7 opened the pull request:
Primarily, this PR adds the manageServices script into the sample directory.  As we got increasing demand to customize VMs as a block, this script is made to try and support customization in a highly extensible way.  

The script is just a wrapper for parallel execution of commands or scripts across large numbers of VMs.

The commands.json file contains the choices and commands, or the location of the script.  To incorporate another command, all anyone needs to do it add it to the commands.json file.  If (s)he wants to deploy a script, just add the script to the action_scripts folder, add the command in json and specify the interpreter to use when launching the script.

Currently supported are:
Enable/Disable Windows Firewall
Enable/Disable SMBv1
Enable/Disable Windows Updates
Enable/Disable Windows Defender (does not work on Windows 1703 or Windows 81, yet)
Enable/Disable Autologin

For future work, it might be nice to leverage the cpe stuff @jmartin-r7 recently added to track supported OSes and methods.

I'd also like to see cases here where we pirate chef/vagrant scripts as well as just doing some cool setup stuff.

Other stuff this does is to add a handy script to help generate vm lists, a sampleLib file with some useful functions, and a lazy way to clear snapshots like we use in testing.

## Example:
Help:
```
python manageServices.py -h
usage: manageServices.py [-h] [-a ACTION] [-k KEYWORD] [-lc]
                         [-sn SNAPSHOTNAME] [-vf VMFILE] [-w WAIT]
                         [-cf CREDSFILE] [-af ACTIONFILE] [-un USERNAME]
                         [-pw PASSWORD] [-t THREADS]
                         hypervisorConfig

positional arguments:
  hypervisorConfig      json hypervisor config

optional arguments:
  -h, --help            show this help message and exit
  -a ACTION, --action ACTION
                        action to run
  -k KEYWORD, --keyword KEYWORD
                        VM search parameter
  -lc, --listCommands   list available commands
  -sn SNAPSHOTNAME, --snapshotName SNAPSHOTNAME
                        name if snapshot to create after changes
  -vf VMFILE, --vmFile VMFILE
                        text file with vms to apply changes
  -w WAIT, --wait WAIT  override the time (in seconds) to wait after running
                        commands/script
  -cf CREDSFILE, --credsFile CREDSFILE
                        credentials file for logging into the vm
  -af ACTIONFILE, --actionFile ACTIONFILE
                        json file containing commands/actions
  -un USERNAME, --username USERNAME
                        vm username
  -pw PASSWORD, --password PASSWORD
                        vm password
  -t THREADS, --threads THREADS
                        concurrent vms to alter

```
List available commands:
```
python manageServices.py -lc x
DEFENDER_SIGUPDATE
UAC_ENABLE
WSUS_DISABLE
DEFENDER_DISABLE
UAC_DISABLE
AUTOLOGIN_ENABLE
SMB1_ENABLE
FIREWALL_DISABLE
SMB1_DISABLE
WSUS_ENABLE
FIREWALL_ENABLE
DEFENDER_ENABLE

```
Turn on Windows firewall for all windows 10 VMs and add a snapshot for that new state called "FIREWALL_ENABLED" updating 5 vms at a time:
```
python manageServices.py -a FIREWALL_ENABLE -sn FIREWALL_ENABLED -k Win10 -un vagrant -pw vagrant -t 5 ../../payload-testing/config/laptop_lab.json 
USING 5 THREADS
('TEST_Win10x86', True)
('TEST_Win10x64_1703', True)
('Win10x64_1703', True)
('Win10x64_1607', True)
('Win10x86_1511', True)
('Win10x86_1703', True)
('Win10x86', True)
('Win10x64_1511', True)
('Win10x86_1607', True)
('Win10x64', True)

```

bwatters-r7 committed:
Added getPassword and cmd encapsulation option for uploadAndRun

bwatters-r7 committed:
fix typo

bwatters-r7 committed:
Ignore the log files produced by sample scipts

bwatters-r7 committed:
Remove duplicate methods

bwatters-r7 committed:
Fix odd return value..... maybe

bwatters-r7 committed:
Add new scripts and a library for vm management

jmartin-tech commented on vm_automation/esxiVm.py:
```diff
@@ -401,6 +401,9 @@ def getVmIp(self):
         return self.vmUsername
 
+    def getPassword(self):
+        return self.vmPassword^M
```
This looks like the `^M` is actually in the line not as  the line terminator?

jmartin-tech commented on samples/sampleLib.py:
```diff
@@ -0,0 +1,58 @@
+
+def waitForProcess(vmObject, procName, timeout = 600):
+    retVal = False
+    for i in range(timeout/5):
```
Check `timeout` for `0` first to avoid attempting to divide by zero.

jmartin-tech commented on samples/sampleLib.py:
```diff
@@ -0,0 +1,58 @@
+
+def makeVmList(vmServer, keywordArg, fileArg):
+    vmList = []
+    if fileArg != None:
```
The arguments here might be better described as `keywordFilter` and `fileFilter`.   Also different workflow comes to mind here.

It might be reasonable to expect the user to read and parse the file first and take arguments as a keyword filter and a known list filter.  If a task preformed was expected to start with a larger list and later wanted to use a subset it could pass in lists of vm names and get the the associated vm objects that exists.

jmartin-tech commented on samples/sampleLib.py:
```diff
@@ -0,0 +1,58 @@
+    return None
+
+def loadJsonFile(fileName):
+    try:
```
Just opinion here but this seems a little more simple.

```
try:
    with open(fileName) as fileObject:
        fileDic = json.load(fileObject)
        return fileDic
except Exception as e:
    print ("UNABLE TO LOAD DICTIONARY FROM: " + str(fileName) + "\n" + str(e))
    return None
```

bwatters-r7 replied:
Honestly, as finicky as the json reader is, I try and maintain distinct errors for file errors and json processing errors.  I know that it will print the error here, but splitting it has always been my preference.

jmartin-tech commented on samples/sampleLib.py:
```diff
@@ -0,0 +1,58 @@
+    else:
+        return False
+
+def getVmObjectFromName(vmServer, vmName):
```
This seems like it might be something to move into `esxiServer` class as `esxiServer.findVmByName(vm_name)` or something of that nature.

bwatters-r7 replied:
I agree

jmartin-tech commented on samples/manageServices.py:
```diff
@@ -0,0 +1,262 @@
+    """
+    logFile = './logs/service_management_' + vmName + '.log'
+    try:
+        os.remove(logFile)
```
This approach to logging does not sit well for me.   Removing log data at the start of a task to me is a poor experience.

The current log file is adding timestamps for each log entry so I don't think the file should be deleted.  It would be reasonable to just add a log line as a marker for a new parallelRun task that is already below.

bwatters-r7 replied:
Not thrilled about it, but I can see your point.  That's probably a less bad way to go.

jmartin-tech commented on samples/manageServices.py:
```diff
@@ -0,0 +1,262 @@
+    """
+    ARE WE USING THREADS?
+    """
+    if args.threads != None:
```
instead of having duplication of the serial case the overhead of a pool with only one executor seems reasonable for the code it removes.

```
num_threads = 1
if args.threads != None:
    num_threads = args.threads
```

bwatters-r7 replied:
I'm OK with that.

jmartin-tech commented on samples/manageServices.py:
```diff
@@ -0,0 +1,262 @@
+import time
+from pyVmomi import vim
+import sampleLib
+from Crypto.SelfTest.Random.test__UserFriendlyRNG import multiprocessing
```
I think this line should be removed.

bwatters-r7 replied:
Ducking auto import.  Good catch.

jmartin-tech commented on samples/action_scripts/commands.json:
```diff
@@ -0,0 +1,252 @@
+							"/f"
+						]
+					]
+		}
```
Should probably offer`AUTOLOGIN_DISABLE` action as well since we are here.

bwatters-r7 replied:
Sure; I thought I added it, but must have forgotten

jmartin-tech commented on .gitignore:
```diff
@@ -4,4 +4,6 @@
 /*.egg-info
 /.idea
 /*.log
+/*.log
```
Remove extra exclude.

bwatters-r7 replied:
Huh; good catch

jmartin-tech reviewed (commented):
Mostly minor changes and suggestions.

jmartin-tech commented on samples/manageServices.py:
```diff
@@ -0,0 +1,262 @@
+        RUN THE COMMANDS IN SERIAL
+        """
+        for vm in vmsToChange:
+            if credsDictionary != None and vm.vmName in credsDictionary:
```
Be sure to incorporate this into the threaded case.

jmartin-tech committed:
PR review comments

Updates make some python standard style updates, progress reporting
and updates to cred file format.
